### PR TITLE
Checks maximum path length using .NET features.

### PR DESCRIPTION
**Pull request by avengerx**
icsharpcode/ILSpy#2706.

### Problem
Okay, this is my best shot. I have heavily simplified `CleanUpName()` method, sacrificing the file truncation functionality in favor of the much more relaxed file name length.

I have tested the edge cases, but by using `new DirectoryInfo(path).FullName` basically I enable any long path support .NET might like, as I could create the files with much longer than 260 characters path during my tests. The problems turns in limitations with Windows applications like Visual Studio, Notepad++, etc. Surprisingly though, puny stock Windows Notepad 
 (`%windir%\system32\notepad.exe`) could read files with very long paths without issues.

Adding the path limit to 258 characters allowed all applications I have to read the generated files (I tested the longest file path the project generated of course), such that if I increased the limit to 259 the applications won't be able to (Visual Studio 2022 actually could, but no longer if I changed it to 260).

I tried also to make very descriptive exceptions when the limit is reached to help users identify it and report if this ever gets accepted mainstream.

### Solution

- rely in `DirectoryInfo` and `FileInfo` classes to validate paths in every write operation
- raise windows hardcoded limits to the absolute maximum usual applications would work with
- no longer truncate -- throw descriptive exceptions pointing the attempted path and its actual length
- use absolute paths on all file operations (expand `TargetDirectory`)

### Disclaimer

This works for me. If you don't fancy the feature or wasting more time on it, feel free to close the pull request on spot.

### CODE REVIEW

File: ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs [hunk 1]
```diff
@@ -581,115 +581,48 @@ public static string CleanUpFileName(string text)
...
 		/// <summary>
 		/// Cleans up a node name for use as a file system name. If <paramref name="separateAtDots"/> is active,
-		/// dots are seen as segment separators. Each segment is limited to maxSegmentLength characters.
-		/// (see <see cref="GetLongPathSupport"/>) If <paramref name="treatAsFileName"/> is active,
-		/// we check for file a extension and try to preserve it, if it's valid.
+		/// dots are seen as segment separators. If <paramref name="lookForExtension"/> is active,
+		/// we check for file a extension and try to preserve it when separating paths at dots.
 		/// </summary>
-		static string CleanUpName(string text, bool separateAtDots, bool treatAsFileName)
+		static string CleanUpName(string name, bool separateAtDots, bool lookForExtension)
```
Comment (siegfriedpammer) [inline ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: You have modified `CleanUpName` to the point, where it no longer fulfills its actual main responsibility. First and foremost, it was meant to **clean up** names (hence the name) used in managed assemblies (which can be arbitrary strings) so that all characters that are not human readable are replaced by '-'.
We added the trimming functionality, because we wanted to avoid having to scan the string multiple times, when creating filenames for resource or type names.

Comment (dgrunwald) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: Note that this cleaning functionality was *security critical*.
Without it is was possible that weird enough class names (e.g. involving `\..\`) cause ILSpy to write outside the output directory.

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: I will review it to restore its functionality. May we at least get rid of the trimming? Make it selectable maybe? This is making embedded resources most likely broken when trimming happen to them.

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: The `new DirectoryInfo` (or `FileInfo`) introduced with `ValidatePath()` nicely blocks any leftover invalid character in path from making it to an actual write. I believe all that's left as far as path _validation_ is concerned, is to add a

```csharp
if (!dotnetPath.StartsWith(TargetDirectory))
  throw new Exception("Path resolved outside of output directory: " + path + Environment.NewLine + "Resolved to: " + dotnetPath);
```

To the validation, ensuring the path is within `TargetDirectory`. At that point, any `../` or relative paths should have already been expanded by `FileSystemInfo`. This would ultimately avoid accidents, right? Now, if a file file has a backtick or dollar sign, it would be another problem (if somebody processes the output files and does not properly treat it as a string, for instance). 

As for the cleanup, something I'm worried about is that characters like `]`, `{`, and many others, including accent, UTF-8, should be accepted as path strings. For instance, the cyrillic `выход` is a valid path name that, as far as I can see, the current implementation of `CleanUp()` wouldn't allow. I want to think a way to effectively allow valid characters go, and clean up only what's clearly valid in managed assembly handles but not in paths. I.e. newlines, tabs, and other control characters.

Comment (siegfriedpammer) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: > The new DirectoryInfo (or FileInfo) introduced with ValidatePath() nicely blocks any leftover invalid character in path from making it to an actual write.

IIRC DIrectoryInfo and friends will do *a lot* more than just cleaning up the path and in order to accomplish that will use disc IO, which we **don't want** at all in that method, as it is called in many places, for example the ILSpy UI for TreeNode labels, not just the project decompiler.

> As for the cleanup, something I'm worried about is that characters like ], {, and many others, including accent, UTF-8, should be accepted as path strings. For instance, the cyrillic выход is a valid path name that, as far as I can see, the current implementation of CleanUp() wouldn't allow.

Then you need to take another careful look: The original implementation uses [`char.IsLetterOrDigit`](https://docs.microsoft.com/en-us/dotnet/api/system.char.isletterordigit?view=net-6.0) (see line 635 below),

https://github.com/icsharpcode/ILSpy/blob/107a5b50442a098e98e62068c29efd32c7b28c2e/ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs#L630-L636

which follows the Unicode Standard and thus supports all characters that are widely in use. Also, .NET Strings are UTF-16 in memory anyway, so UTF-8 is of no concern at this point. Any string that can be stored in the ECMA-335 metadata tables, may be either UTF-8 or UTF-16 and is supported by .NET's built-in functions.

```csharp
> "выход".All(char.IsLetterOrDigit)
true
> "출구".All(char.IsLetterOrDigit)
true
> "saída".All(char.IsLetterOrDigit)
true
```

You see, **it properly supports** Cyrillic script, Hangul (Korean) script and Portuguese (see the acute accent on the i?) and I could go on with Japanese, Chinese and many others. Thank you!

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: > IIRC DIrectoryInfo and friends will do a lot more than just cleaning up the path and in order to accomplish that will use disc IO, which we don't want at all in that method, as it is called in many places, not just the project decompiler.

You are right, and as I'm developing here I'm seeing the alternative _without_ using `DirectoryInfo` _except_ when we really want to write a file. I'll keep that in mind, so thanks for the heads up.

> Then you need to take another careful look: The original implementation uses [char.IsLetterOrDigit](https://docs.microsoft.com/en-us/dotnet/api/system.char.isletterordigit?view=net-6.0) (see line 635 below),

Sorry I really made you angry here again... My argument was with this in mind:

```
Console.WriteLine("Is bracket valid: " + char.IsLetterOrDigit('['));
Is bracket valid: False
```

So instead, I'm thinking in looking up _what is not valid_ instead of _what is valid_ in the cleanup. I guess it would work best if I show you the code when it's done. What's your position about regular expressions at that point?

fun fact: this came to mind because I recently noticed I have installed Linqpad in `c:\apps\linqpad5]`, and it's been working like a charm!

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: @siegfriedpammer what's the plot behind [this positioning after colon](https://github.com/icsharpcode/ILSpy/blob/e2d79774cd3eb98b2b14f2a1fbeda1d9d0089d60/ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs#L595-L597)? Just our usual `c:\path`? Or is there any other situation where a node will have a colon and you'd want to skip the string until it? Any that comes on top of your head so I can ensure the case is safe?

Comment (siegfriedpammer) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: > Sorry I really made you angry here again... 

I am sorry for being unfair to you. I can see your good intent, but so far your suggestions seem to be aimed only at solving your specific problem, while introducing a lot of other problems/bugs in other cases at the same time. I am sure, we will find a satisfying solution for your problem as well.

> So instead, I'm thinking in looking up what is not valid instead of what is valid in the cleanup.

Sorry, but "whitelisting" (only allowing a defined set of things) is considered more secure than "blacklisting" (only removing a defined set of things). As @dgrunwald already pointed out, this code must be as secure as possible, because it is used when writing files. Many of our users are security researchers, who are analyzing .NET malware with ILSpy. We need to make sure that there is absolutely no unexpected behavior when working with ILSpy. And only a allowing a "known good/harmless set of characters" is the best approach with regards to security.

Regarding brackets: Just like we allow "_" in names, we could add other valid characters to the list as well, however, it quickly will get very complicated, because there are different file systems that have different restrictions (see also https://en.wikipedia.org/wiki/Filename#Reserved_characters_and_words for more details).

We want our code to be as portable as possible, hence we have a very limited list of characters we allow in filenames.

> What's your position about regular expressions at that point?

Why use regular expressions when a simple O(n) scan is sufficient?

> what's the plot behind [this positioning after colon](https://github.com/icsharpcode/ILSpy/blob/e2d79774cd3eb98b2b14f2a1fbeda1d9d0089d60/ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs#L595-L597)?

`:` is only valid in Windows paths, when the path is an absolute path, `:` is the second character and the first character is `A-Za-z`. It can only be used to denote a "drive-letter" at start of an absolute path.

```
❯ mkdir ab:
The filename, directory name, or volume label syntax is incorrect.
```

But you would have to ask @dgrunwald why we are stripping that part completely instead of replacing it with '-'.

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: Okay, I see... Let me start this from scratch once again :rofl: now with a more conservative approach. And with clear goals:
- allow file name characters over 30 in length on Windows without long paths enabled
- check full path length only when I'm about to write files (thus properly checking full path limits when we actually have that information)

So perhaps from the next attempt we could work towards having the best of security without sacrificing path lengths, thus increasing the odds of the ability to extract embedded resources that can actually match references in the code, given the constraints that will always be there.

At least by the comprehensiveness of your responses I assume this is something you'd also want to see addressed, right?

Comment (dgrunwald) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: "check full path length only when I'm about to write files" -- I'm not sure why you feel the need to check this yourself when the operating system will already do so for us. 
If you don't use the total path length for any sort of truncation then there isn't at any point in checking it, is there?

Comment (dgrunwald) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: Assuming that .NET 6 can write long paths even if the registry is key is not set, then I would suggest just removing the registry check and proceeding as if long path support is enabled.

Comment (avengerx) [reply ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs:592]: > "check full path length only when I'm about to write files" -- I'm not sure why you feel the need to check this yourself when the operating system will already do so for us. If you don't use the total path length for any sort of truncation then there isn't at any point in checking it, is there?

Reasons I feel the need:
1. produce more meaningful error messages for the user, allowing them to know what exact path is breaking and by how many characters
2. avoid an issue where files can be created yet most Windows applications won't open them (with "cannot find part of path" much like #2622 initial report)

In my tests, simply letting the OS tell the length (at least on Windows 10, no long paths enabled), files generated on very long paths are able to be opened with Windows Notepad (%windir%\System32\notepad.exe).

At least Visual Studio 2022, WordPad, and Notepad++, they won't open these edge cases:
- VS2022 tells it exceeds 259 chars length,
- notepad++ silently ignores it
- WordPad closes on spot

> Assuming that .NET 6 can write long paths even if the registry is key is not set, then I would suggest just removing the registry check and always proceeding as if long path support is enabled.

It may be the case and perhaps I'm a bit lost here. The project I'm fiddling with [says it is `netstandard2.0`](https://github.com/icsharpcode/ILSpy/blob/e2d79774cd3eb98b2b14f2a1fbeda1d9d0089d60/ICSharpCode.Decompiler/ICSharpCode.Decompiler.csproj#L5),

**Comment by avengerx** [comment]
Current state already passes all unit tests. Making validation specific unit tests. It's a bit tricky to go multi-platform and all... :P

**Comment by dgrunwald** [comment]
I think this PR is way too complex. We should be able to simplify the code thanks to .NET 6. This PR is the opposite of that. We're not going to accept this approach.
